modules/internal/ChapelIO.chpl

```diff
@@ -345,50 +349,225 @@ module ChapelIO {
-          } else {
-            if !isBinary {
+        if isBinary {
+          for param i in 1..num_fields {
+            if isType(__primitive("field value by num", x, i)) ||
+               isParam(__primitive("field value by num", x, i)) {
+              // do nothing, don't read types or params
+            } else {
+              reader.readwrite(__primitive("field value by num", x, i));
+            }
+          }
+        } else if num_fields > 0 {
+          // track whether or not we've read all of the fields.
+
+          // this tuple helps us not read the same field twice.
+          var read_field:(num_fields)*bool;
+          var num_to_read = 0;
+          var num_read = 0;
+          for param i in 1..num_fields {
+            if isType(__primitive("field value by num", x, i)) ||
+               isParam(__primitive("field value by num", x, i)) {
+              // do nothing, don't read types or params
+            } else {
+              num_to_read += 1;
+            }
+          }
+
+          // the order should not matter.
+          while num_read < num_to_read {
+            //writeln("IN LOOP read_field=", read_field);
+
+            if needsComma {
+              // read a comma
+
+              //writeln("READING COMMA");
               var comma = new ioLiteral(",", true);
-              if !first then reader.readwrite(comma);
-    
-              var fname = new ioLiteral(__primitive("field num to name", t, i), true);
-              reader.readwrite(fname);
-    
-              var st = reader.styleElement(QIO_STYLE_ELEMENT_AGGREGATE);
-              var eq:ioLiteral;
-              if st == QIO_AGGREGATE_FORMAT_JSON {
-                eq = new ioLiteral(":", true);
+              reader.readwrite(comma);
+
+              if !reader.error() {
+                needsComma = false; // we read a comma
+              } else if reader.error() == EFORMAT {
+                //writeln("NO COMMA");
+                // break out of the loop if we didn't read a comma
+                // and we're expecting to read one.
+                break;
+              }
+            }
+
+            // find a field name that matches.
+            // TODO: this is not particularly efficient. If we
+            // have a lot of fields, this is O(n**2), and there
+            // are other potential problems with string reallocation.
+            var st = reader.styleElement(QIO_STYLE_ELEMENT_AGGREGATE);
+            var skip_unk = reader.styleElement(QIO_STYLE_ELEMENT_SKIP_UNKNOWN_FIELDS);
+
+            // the field name
+            var fname:ioLiteral;
+
+            var read_field_name = false;
+
+            //writeln("BEFORE PARAM FOR type ", t:string, " num_fields=", num_fields, " read_field_name=", read_field_name);
+
+            for param i in 1..num_fields {
+              if isType(__primitive("field value by num", x, i)) ||
+                 isParam(__primitive("field value by num", x, i)) {
```

Comment:
It might be worthwhile to make this pair of checks into a single call so you can reuse it.